### PR TITLE
[Programming]Add optimize attribute checker

### PR DESCRIPTION
**Pull request by sneaxiy**
Support attributes `worker.*`, `data.*` and `solver.*`.

### CODE REVIEW

**Review by typhoonzero — DISMISSED** [review]

File: pkg/sql/codegen/optimize/template_optimize.go [hunk 1]
```diff
@@ -75,27 +83,36 @@ def submit():
...
+                       batch_size=batch_size)
     
     input_conf = InputConf(df1=df1)
 
     optflow_version = os.environ.get("SQLFLOW_OPTFLOW_VERSION")
     if not optflow_version:
         raise ValueError("Environment variable SQLFLOW_OPTFLOW_VERSION must be set")
     	
-    if "{{.IsPAI}}":
+    if is_pai:
```
Comment (typhoonzero) [inline pkg/sql/codegen/optimize/template_optimize.go]: check worker count > 1. `is_pai` should not be used in optimize

Comment (sneaxiy) [reply pkg/sql/codegen/optimize/template_optimize.go]: As offline discussion, `is_pai` is removed. We can only support kubemaker mode.

File: pkg/sql/codegen/optimize/codegen.go [hunk 2]
```diff
@@ -15,21 +15,84 @@ package optimize
...
+	"encoding/json"
 	"fmt"
 	"sqlflow.org/sqlflow/pkg/ir"
 	pb "sqlflow.org/sqlflow/pkg/proto"
+	"sqlflow.org/sqlflow/pkg/sql/codegen/attribute"
 	"strings"
 	"text/template"
 )
 
+func checkIsPositiveInteger(i interface{}, name string) error {
```
Comment (wangkuiyi) [inline pkg/sql/codegen/optimize/codegen.go]: Move `checkIsPositiveInteger` into package `attribute.IsPositiveInteger`?

File: pkg/sql/codegen/optimize/codegen.go [hunk 3]
```diff
@@ -15,21 +15,84 @@ package optimize
...
+	if i.(int) <= 0 {
```
Comment (wangkuiyi) [inline pkg/sql/codegen/optimize/codegen.go]:
```go
if v, ok := i.(int); !ok || v <= 0 {
```
 
Check https://tour.golang.org/methods/15 for more examples.

File: pkg/sql/codegen/optimize/codegen.go [hunk 4]
```diff
@@ -15,21 +15,84 @@ package optimize
...
+		return checkIsPositiveInteger(i, "worker.num")
+	}},
+	"worker.core": {attribute.Int, 8, "Worker core number", func(i interface{}) error {
+		return checkIsPositiveInteger(i, "worker.core")
+	}},
+	"worker.memory": {attribute.Int, 4096, "Worker memory", func(i interface{}) error {
+		return checkIsPositiveInteger(i, "worker.memory")
+	}},
+	"solver.*": {attribute.Unknown, nil, "Solver options", nil},
+}
```
Comment (wangkuiyi) [inline pkg/sql/codegen/optimize/codegen.go]: The Go idiomatic way is

```go
var attributeDict = attribute.NewDict().
   Bool("data.enable_slice", false, "the explanation", nil).
   Int32("worker.core", 8, "number of cores per worker", attribute.PositiveInt32)
```

Then, we don't need to abuse `interface {}` in `attribute.PostiveInterger`

```go
// In package attribute.assert.
func PositiveInt32(x int, name string) error {
    if x <= 0 {
        return fmt.Errorf("Expecting %s being positive integer, but got %d", name, x)
    }
    return nil
}
```

File: pkg/sql/codegen/optimize/codegen.go [hunk 5]
```diff
@@ -15,21 +15,84 @@ package optimize
...
+
+const (
+	dataAttrPrefix   = "data."
+	solverAttrPrefix = "solver."
+	workerAttrPrefix = "worker."
```
Comment (wangkuiyi) [inline pkg/sql/codegen/optimize/codegen.go]: Either make sure that the construction of the `attribute.Dictionary` uses these prefixes too, or, if they are only used in `func GenerateOptflowOptimzieCode`, move them into that function definition.

File: pkg/sql/codegen/optimize/template_optimize.go [hunk 6]
```diff
@@ -57,16 +57,20 @@ class CustomOptFlowRunner(BaseOptFlowRunner):
 
 const optFlowSubmitText = `
 import os
+import json
```
Comment (wangkuiyi) [inline pkg/sql/codegen/optimize/template_optimize.go]: It doesn't make sense to have a template this long -- please file a PR to minimize the template into only one function call which takes all template-filler parameters, before merging this one.

**Review by wangkuiyi — CHANGES_REQUESTED** [review]

File: pkg/sql/codegen/optimize/codegen.go (hunk 3 above)
Comment (sneaxiy) [reply pkg/sql/codegen/optimize/codegen.go]: Done.

File: pkg/sql/codegen/optimize/codegen.go (hunk 5 above)
Comment (sneaxiy) [reply pkg/sql/codegen/optimize/codegen.go]: Done.

File: pkg/sql/codegen/optimize/template_optimize.go (hunk 6 above)
Comment (sneaxiy) [reply pkg/sql/codegen/optimize/template_optimize.go]: Done. Move most Python code to `runner.py`.

**Review by wangkuiyi — APPROVED** [review]

**Pull request merged**